# Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 0391 Кононенко Евгений, вариант 57.

## Цель работы

Цель работы приводится в соответствии с указаниями к практической работе.

## Основные теоретические положения

$\quad$ Аддитивная цепочка - последовательность натуральных чисел от 1 до n
$$ 1 = a_0, a_1, ..., a_r = n ,$$ 
где каждый элемент равен сумме каких-либо двух предыдущих элементов
$$ a_{i} = a_{j} + a_{k}, k \leq j < i $$

$ l(n) $ - наименьшая длина аддитивной цепочки для заданного $ n \in \mathbb{N} $.

Неравенство для метода множителей: $ \quad l(mn) \leq l(m) + l(n) $

Неравенство для m-арного метода $ [m = 2^k, n = \sum_{j=0}^{k} d_{j}m^{t-j}] $: $ \quad l(n) = m - 2 + (k+1)t $

Неравенство для бинарного алгоритма "SX": $ \quad l(n) = \lambda(n) + \nu(n) - 1 $

### Свойства аддитивных цепочек
* Полагается, что все аддитивные цепочки - возрастающие:
$ 1 = a_0 < a_1 < ... < a_r = n $
* Если два числа из последовательности $\{a_i\}$ совпадают, то одно из них может быть опущено
* Пара $ (j, k), 0 \leq k < j < i $ называется шагом
* Если существует более одной пары $ (j, k) $, полагается, что $ j $ - наибольшее из возможных 

### Виды шагов 
* удвоение: $ \quad j = k = i - 1 $
* звездный шаг: $ \quad j = i - 1 $
* малый шаг: $ \quad \lambda(a_{i}) = \lambda(a_{i-1}) $

### Свойства шагов
* Первый шаг - всегда удвоение
* Удвоение - звездный шаг, но никогда не малый шаг
* За удвоением всегда следует звездный шаг
* Если $ i-й $ шаг не малый, то $ (i + 1)-й $ - либо звездный, либо малый либо и тот, и другой
* Если $ (i + 1)-й $ шаг ни звездный, ни малый, то $ i-й $ шаг должен быть малым

### Теорема
Если аддитивная цепочка включает $ d $ удвоений и $ f = r - d $ неудвоений, то 
$$ n \leq 2^{d-1}F_{f+3}, $$
где $ F_j $ - число Фибоначчи под номером $ j $.
### Следствие из теоремы
Если аддитивная цепочка содержит $ f $ удвоений и $ S $ малых шагов, то
$$ S \leq f \leq \frac{S}{1-log_2(\varphi)}, $$ 
где $ \varphi = \frac{\sqrt{5} + 1}{2} $ - золотое сечение.

### Алгоритм Брауэра
Для заданного числа $ n \in \mathbb{N} $ можно построить цепочку Брауэра с помощью рекуррентной формулы
$$ B_k(n) = 
\begin{cases} 
1, 2, 3, ..., 2^{k} - 1,\  если \  n < 2^{k}\\
B_k(q), 2q, ..., 2^kq, n,\  если \  n \geq 2^{k} и \  q = \lfloor{\frac{n}{2^k}} \rfloor
\end{cases} $$

Данная цепочка будет иметь длину
$$ l_B(n) = j(k + 1) + 2^k - 2 $$
при условии, что $ jk \leq \lambda(n) < (j + 1)k $

Длина цепочки будет минимизирована для больших $ n $, если 
$$ k = \lambda(\lambda(n)) - 2\lambda(\lambda(\lambda(n))) $$

#### Уточнения
* Задается фикированный параметр $ k $. Выполняются вычисления вспомогательных чисел:
$$ 
d = 2^k, q_1 = [\frac{n}{d}], r_1 = n \ mod \ d \Rightarrow n = q_{1}d + r_{1} \ (0 \leq r_{1} < d) \\
q_2 = [\frac{q_1}{d}], r_2 = q_1 \ mod \ d \Rightarrow q_1 = q_{2}d + r_{2}
$$
* Данная процедура выполняется до тех пор, пока $ q_s $ не станет меньше $ d $. Следовательно $ q_{s-1} = q_{s}d + r_{s} $
* Таким образом,
$$ n = 2^kq_1 + r_1 = ... = 2^k(2^k(...(2^kq_s + r_s)...) + r_2) + r_1 $$
$$ B_k(n): 1, 2, 3, ..., 2^k - 1, 2q_s, 4q_s, ..., 2^kq_s + r_s, ..., n $$

### Алгоритм Яо
* Имеет ту же вычислительную сложность, что и алгоритм Брауэра
* Раскладываем число $ n \in \mathbb{N} $ в $ 2^k-ой $ системе счисления:
$$ n = \sum_{i=0}^j a_{j}2^{ik}, \ a_j \neq 0 $$
* Введем функцию $ d(z) $:
$$ d(z) = \sum_{i: a_i = z} 2^{ik} $$

#### Ход алгоритма
* Задаем базовую последовательность: $ 1, 2, 4, ..., 2^{\lambda(n)} $
* Вычисляем значения $ d(z) $ для $ z \in \{ 1, 2, 3, ..., 2^k - 1 \} $, причем $ d(z) \neq 0 $
* Вычисляем $ zd(z) $ для всех $ z $ 
* В конечном итоге получаем
$$ n = \sum_{z = 1}^{2^k - 1} zd(z) $$

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать.


## Порядок выполнения работы

1. Вручную (т.е. не реализовывая алгоритм на Sage) построить последовательность вычислений бинарным методом и методом множителей для \\[x^n\\] для 2-3 значений 𝑛 (значение 𝑛 > 30 выбираются студентом самостоятельно). Сравнить количество операций для каждого метода, сделать выводы.


__**Бинарный метод**__


\\[l(n)= \lambda(n)+ \nu(n) - 1,\\]
\\[\text{где } \lambda(n)- \text{уменьшенная на единицу длина бинарной записи числа n, }\nu(n) - \text{количество единиц в бинарной записи числа n.}\\]

По этой формуле мы можем посчитать количество операций бинарным методом.

__Для n = 33__ мы имеем **6** операций

| i | N  | Y    | Z    |
|---|----|------|------|
| 0 | 33 | 1    | x    |
| 1 | 16 | x    | x^2  |
| 2 | 8  | x    | x^4  |
| 3 | 4  | x    | x^8  |
| 4 | 2  | x    | x^16 |
| 5 | 1  | x    | x^32 |
| 6 | 0  | x^33 | x^32 |


__Для n = 43__ мы также имеем **8** операций

| i | N  | Y    | Z    |
|---|----|------|------|
| 0 | 43 | 1    | x    |
| 1 | 21 | x    | x^2  |
| 2 | 10 | x^3  | x^4  |
| 3 | 5  | x^3  | x^8  |
| 4 | 2  | x^11 | x^16 |
| 5 | 1  | x^11 | x^32 |
| 6 | 0  | x^43 | x^32 |

__Для n = 57__ мы также имеем **8** операций

| i | N  | Y    | Z    |
|---|----|------|------|
| 0 | 57 | 1    | x    |
| 1 | 28 | x    | x^2  |
| 2 | 14 | x    | x^4  |
| 3 | 7  | x    | x^8  |
| 4 | 3  | x^9  | x^16 |
| 5 | 1  | x^25 | x^32 |
| 6 | 0  | x^57 | x^32 |


__**Метод множителей**__

__Для n = 33__ мы имеем **7** операций

\\[x^{33}\\]
\\[\text{33 = 11*3} \\]
\\[\text{Пусть х^3 = y : тогда x,x^2,x^3 - 2 операции}\\]
\\[\text{11 = 1+10} \\]
\\[y^{11} = y^{10}*y=((y^5)^2)=(((y^2)^2*y)^{2}*y) \text{   - 5 операций}\\]

Итого имеем 5+2=**7** операций - и проигрываем бинарному методу

__Для n = 43__ мы имеем **7** операций
\\[x^{43}\\]
\\[\text{43 = 42+1 - 1 операция } \\]
\\[\text{Пусть х^2 = y : тогда x,x^2 - 1 операция}\\]
\\[y^{24} = (y^{2})^{12} = (((y^{2}*y)^{2})^{2})^{2} \text{   - 5 операций}\\]
Итого имеем 5+1=**7** операций - и выигрываем бинарный метод

__Для n = 57__ мы имеем **8** операций
\\[x^{57}\\]
\\[\text{57 = 56+1 - 1 операция} \\]
\\[x^{56} = ((x^{2})^{28}) = (((x^{2})^{2})^{2})^{7} \text{ }\\]
\\[\text{Пусть (((x^2)^2)^2) = y - 3 операции}\\]
\\[\text{7 = 6 + 1 } \\]
\\[y^{7} = y*(y^{2})^{3} = y*(y^{2}*y)^{2} \text{   - 4 операции}\\]
Итого имеем 4+3+1=**8** операций - точно также как и бинарным методом


2. Реализовать алгоритм Брауэра для вычисления приближённых аддитивных цепочек для различных чисел при варьировании параметра 𝑘, сопоставить длины полученных аддитивных цепочек с минимальной аддитивной цепочкой для заданного числа. Сделать выводы.



In [1]:
#Импортируем floor для деления с округлением в меньшую сторону
from math import floor
#Вычислим длину бинарного кода - 1
def lmbda(n):
    return n.nbits() - 1
#Рассчет теоретического k
def calculK(n):
    return lmbda(lmbda(n))-2*lmbda(lmbda(lmbda(n)))
#Рассчет теоретической длины цепочки
def minLenChain(n):
    return int(lmbda(n) + lmbda(n) / lmbda(lmbda(n)) + (lmbda(n) * lmbda(lmbda(lmbda(n)))) / (lmbda(lmbda(n)) ** 2))

def brauer(n,k):

#Создаем два массива: qArr - для хранения результатов операции floor(n/d), т.е. неполных частных, 
                    # rArr - для хранения результатов операции n % d, т.е. остатков при делении
    qArr=[]
    rArr=[]
    # И конечно же сам делитель числа n
    d = pow(2,k)
    # Массив результатов - она и аддитивная цепочка
    additiveChain = []

    # Изначально массив заполнен до 1,2,3,...,2^(k)-1
    for i in range(1,d):
        additiveChain.append(i)

    # Записываем в массивы все q и r, важно понимать, цикл не бесконечен,  
                                       # т.к. мы уменьшаем n: n = floor(n/d), когда то это значение обязательно станет нулевым 
    while(d < n):

        qArr.append(floor(n / d))
        rArr.append(n % d)
        n = floor(n / d)


    #Отзеркаливаем массив значений q, для дальнейшего удобства
    qArr = qArr[::-1]

    #Добавляем значения в массив - достраиваем аддитивную цепочку
    for q in qArr:
        for i in range(1,k+1):
            additiveChain.append(pow(2,i)*q)
        additiveChain.append(additiveChain[-1] + rArr[-1])
        rArr.pop(-1)
    return sorted(set(additiveChain))

k = 3
n = 31415
adChain = brauer(n,k)
lenMinChain = minLenChain(n)
minK = calculK(n)
print("Аддитивная цепочка:", *adChain)
print("Теоретическая длина аддитивной цепочки:", lenMinChain)
print("Длина аддитивной цепочки:", len(adChain))
print("Теоретически минимальное k:", minK)

Аддитивная цепочка: 1 2 3 4 5 6 7 14 28 56 61 122 244 488 490 980 1960 3920 3926 7852 15704 31408 31415
Теоретическая длина аддитивной цепочки: 20
Длина аддитивной цепочки: 23
Теоретически минимальное k: 1



Аддитивная цепочка: 1 2 3 4 5 6 7 14 28 56 61 122 244 488 490 980 1960 3920 3926 7852 15704 31408 31415

Теоретическая длина аддитивной цепочки: 20

Длина аддитивной цепочки: 23

Теоретически минимальное k: 1

Длина цепочки может быть минимизирована для большех n, при взятии такого k: \\[k = \lambda\lambda(n)-2\lambda\lambda\lambda(n) \\]
Для n = 31415 длина аддитивной цепочки минимальна при k = 2,3 
В таблице мы можем увидеть градацию длины цепочки от изменения параметра k

| l(n) | k |
|--------|---|
| 25     | 1 |
| 23     | 2 |
| 23     | 3 |
| 29     | 4 |
| 43     | 5 |
| 74     | 6 |
| 137    | 7 |

Из таблицы видно, что длина цепочки минимальна при k = 2, 3, однако проверив по формуле мы получаем k = 1. Как ни странно на практике и длина цепочки вышла чуть больше, чем теоретически должна была получится. Мне кажется, это связано с тем что мы взяли достаточно не большой n, тем самым мы увидели недостаточное теоретическое представление минимизации k и длины цепочки.

3. Реализовать алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа. Протестировать алгоритм минимум для 5 значений 𝑛 > 1000. Указать, сколько времени потребовалось на поиск цепочки и какая цепочка получилась. Сравнить с предыдущими методами, сделать выводы

In [56]:
from math import ceil,log2,floor
from time import time


def lmbda(n):
    return n.nbits() - 1

def nu(n): 
    return bin(n).count('1')

def getAq(vec):
    chain = [1]
    for i in range(len(vec)):
        chain.append(chain[i] + chain[vec[i] - 1])
    return chain[len(chain) - 1]    

 

def fillParts(l,firstItem):
    part = [0]*l
    for i in range(l):
        part[i] = firstItem + i
    return part    

def isNilVec(vec):
    for i in vec:
        if i > 1:
            return False
    return True

def calc_from_index(v):
    answer = []
    answer.append(1)
    for el in v:
        answer.append(answer[-1] + answer[el-1])
    return answer

def decreaseVec(vec):
    l = len(vec)
    for i in range(l):
        if vec[l - i - 1] > 1:
            vec[l - i - 1] -= 1   
        for j in range(l - i, l):
            vec[j] = j + 1
        break

def crushIndVect(n):
    lMin = ceil(log2(n))
    lMax = lmbda(n) + nu(n)
    vec=[]
    fixedPart = []
    changePart = []
    for m in range(lMin,lMax):  
        for i in range(1,m + 1):
            vec.append(i)
            
       
        q = int(m / 2)
        fixedPart = fillParts(q , 1)
        
        
        while True:
            
            
            changePart = fillParts(m - q,q + 1)


            a = getAq(fixedPart)
            aMin = a + m - q
            aMax = a * 2 ** (m - q)


            if n < aMin or n > aMax:
                if isNilVec(fixedPart) and len(fixedPart) > 1:
                    break
                decreaseVec(fixedPart)
                continue


            while True:
                
                if getAq(fixedPart + changePart) == n:
                    return fixedPart + changePart

                if isNilVec(changePart):
                    break

                decreaseVec(changePart)

            if isNilVec(changePart):
                break

            decreaseVec(changePart)
            
            
def getChain(vec):
    chain = [1]
    for i in range(len(vec)):
        chain.append(chain[i] + chain[vec[i] - 1])
    return chain
    
    
            
    
n = 1025

startTime = time()
vec = crushIndVect(n)
print(vec)
endTime = time() - startTime
print(endTime)
chain = getChain(vec) 
print(chain)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 1]
0.0455479621887207
[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1025]


Проверим алгоритм для 5-ти n:

| time,sec | n    | lenght |
|----------|------|--------|
| 5.2451   | 1000 | 13     |
| 0.00116  | 1024 | 11     |
| 0.00263  | 1025 | 12     |
| 0.44211  | 1033 | 13     |
| 0.00271  | 1010 | 14     |

Пожалуй, очевидно, что чем ближе число к степени двойки, тем оно будет рассчитываться быстрее.
Можно сравнить результаты с алгоритмом Брауэра.


| time,sec | n    | lenght |
|----------|------|--------|
| 5.2451   | 1000 | 13     |
| 0.00116  | 1024 | 11     |
| 0.00263  | 1025 | 12     |
| 0.44211  | 1033 | 13     |
| 0.00271  | 1010 | 14     |

Алгоритмы дают нам очень похожие результаты, а то и точно такие же, но алгоритм дробления вектора индексов работает в разы больше(для чисел отдаленных от степеней 2), из-за того, что он основан на практическом переборе. Для чисел немного меньших степеней двойки результатов можно ждать несколько минут, а то и часов. 

In [ ]:
for n in range(1,12 + 1):
    a = Integer(n)
    b = Integer(2 ** n - 1)
    chainA = crushIndVect(a)
    chainB = crushIndVect(b)
    lenA = len(chainA)
    lenB = len(chainB)
    print(f"n={n}, l(n) = {lenA}, l(2^n-1) = {lenB}")
    print(f"l(2^n - 1) <= l(n) + n - 1: {lenB <= n - 1 + lenA}\n")
    print()

n=1, l(n) = 0, l(2^n-1) = 0
l(2^n - 1) <= l(n) + n - 1: True


n=2, l(n) = 1, l(2^n-1) = 2
l(2^n - 1) <= l(n) + n - 1: True




n=1, l(n) = 1, l(2^n-1) = 1 l(2^n - 1) <= l(n) + n - 1: True

n=2, l(n) = 2, l(2^n-1) = 3 l(2^n - 1) <= l(n) + n - 1: True

n=3, l(n) = 3, l(2^n-1) = 5 l(2^n - 1) <= l(n) + n - 1: True

n=4, l(n) = 3, l(2^n-1) = 6 l(2^n - 1) <= l(n) + n - 1: True

n=5, l(n) = 4, l(2^n-1) = 8 l(2^n - 1) <= l(n) + n - 1: True

n=6, l(n) = 4, l(2^n-1) = 9 l(2^n - 1) <= l(n) + n - 1: True

n=7, l(n) = 5, l(2^n-1) = 11 l(2^n - 1) <= l(n) + n - 1: True

n=8, l(n) = 4, l(2^n-1) = 11 l(2^n - 1) <= l(n) + n - 1: True

n=9, l(n) = 5, l(2^n-1) = 13 l(2^n - 1) <= l(n) + n - 1: True

n=10, l(n) = 5, l(2^n-1) = 14 l(2^n - 1) <= l(n) + n - 1: True

n=11, l(n) = 6, l(2^n-1) = 16 l(2^n - 1) <= l(n) + n - 1: True

n=12, l(n) = 5, l(2^n-1) = 16 l(2^n - 1) <= l(n) + n - 1: True

Гипотеза верна!

## Выводы

Общий вывод по проделанной работе.

Вывод: в ходе выполнения данной практической работы мы научились реализовывать и применять алгоритмы для нахождения минимальных аддитивных цепочек, мы также выработали навыки использования системы компьютерной математики SageMath для этих алгоритмов. На практике была проверены алгоритмы Брауэра и дробления вектора индексов, и сравнение их. Также удалось проверить гипотезу Шольца-Брауэра, которая, как ни странно, оказалось верна.